src/ctap/data_formats.rs

```diff
@@ -445,66 +445,91 @@ pub struct PublicKeyCredentialSource {
     pub user_handle: Vec<u8>, // not optional, but nullable
     pub other_ui: Option<String>,
     pub cred_random: Option<Vec<u8>>,
+
+    /// Contains the unknown fields when parsing a CBOR value.
+    ///
+    /// Those fields could either be deleted fields from older versions (they should have reserved
+    /// tags) or fields from newer versions (the tags should not be reserved). If this is empty,
+    /// then the parsed credential is probably from the same version (but not necessarily).
+    pub unknown_fields: BTreeMap<cbor::KeyType, cbor::Value>,
+}
+
+// We simulate protocol buffers in CBOR with maps. Each field of a message is associated with a
+// unique tag, implemented with a CBOR unsigned key.
+#[repr(u64)]
+enum PublicKeyCredentialSourceField {
+    CredentialId = 0,
+    PrivateKey = 1,
+    RpId = 2,
+    UserHandle = 3,
+    OtherUi = 4,
+    CredRandom = 5,
+    // When a field is removed, its tag should be reserved and not used for new fields. We document
+    // those reserved tags below.
+    // Reserved tags: none.
+}
+
+impl From<PublicKeyCredentialSourceField> for cbor::KeyType {
+    fn from(field: PublicKeyCredentialSourceField) -> cbor::KeyType {
+        (field as u64).into()
+    }
 }
 
 impl From<PublicKeyCredentialSource> for cbor::Value {
     fn from(credential: PublicKeyCredentialSource) -> cbor::Value {
-        let mut private_key = [0u8; 32];
+        use PublicKeyCredentialSourceField::*;
+        let mut private_key = [0; 32];
         credential.private_key.to_bytes(&mut private_key);
-        let other_ui = match credential.other_ui {
-            None => cbor_null!(),
-            Some(other_ui) => cbor_text!(other_ui),
-        };
-        let cred_random = match credential.cred_random {
-            None => cbor_null!(),
-            Some(cred_random) => cbor_bytes!(cred_random),
-        };
-        cbor_array! {
-            credential.credential_id,
-            private_key,
-            credential.rp_id,
-            credential.user_handle,
-            other_ui,
-            cred_random,
+        cbor_extend_map_options! {
+            credential.unknown_fields,
+            CredentialId => Some(credential.credential_id),
+            PrivateKey => Some(private_key.to_vec()),
+            RpId => Some(credential.rp_id),
+            UserHandle => Some(credential.user_handle),
+            OtherUi => credential.other_ui,
+            CredRandom => credential.cred_random
         }
     }
 }
 
-impl TryFrom<cbor::Value> for PublicKeyCredentialSource {
```

Comment:
All data formats implement the `TryFrom<cbor::Value>` to convert from CBOR values. What's the purpose of implementing a novel `parse_cbor` function instead, which isn't tied to any trait?

Reply:
It's because it looked to me that the others struct are specified by FIDO. In particular which error should be returned for which kind of issue. In the case of credentials, the kind of issue is irrelevant because it is not a user input but something coming from the persistent storage. So all errors should be something like "storage format error". We could as well use `()` as the error type. But it will add a bit of boilerplate to convert all errors to `()`.

And actually it could make sense to want different implementations for the `TryFrom<cbor::Value>` and this `parse_cbor` one. Because the first would be the FIDO version while this one would be the internal representation we use for the persistent storage. We should try to not conflate both since it will introduce future issues.

Reply:
Returning `Ctap2StatusCode::CTAP2_ERR_INVALID_CBOR` and then calling `.ok()` to convert the `Result` to an `Option` would be fine. Otherwise I think that this change introduces to much fragmentation.

Reply:
Before doing the change I want to be sure I understand what you mean. What I understand is:
- We don't expect to have to parse credentials from user input in the future.
- Because of that, this is OK to reuse the same API as user input parsing.
- As such we can use the representation we want because it won't conflate with future use.

Is that correct?

Reply:
> * We don't expect to have to parse credentials from user input in the future.
> * Because of that, this is OK to reuse the same API as user input parsing.
> * As such we can use the representation we want because it won't conflate with future use.

My comments have nothing to do with representation. The representation is the same (some CBOR encoding).

This is simply about having a **consistent parsing API** from CBOR to structs. That is, implementing `impl TryFrom<cbor::Value> for T` instead of having each `T` calling its parsing function however it likes. Otherwise we'll end up with fragmentation between `T::from_cbor`, `T::parse_cbor`, `T::from`, `cbor::Value::into::<T>`, and what not.

If `TryFrom<cbor::Value>` isn't the right API, we can revisit this and use another trait, as long as it fits all data formats in OpenSK. But so far it has worked well.

In any case, `PublicKeyCredentialSource` was already implementing `TryFrom<cbor::Value>`, so changing this API is independent from the core change of this pull request (using a map instead of an array).

So this pull request should focus on addressing one change (the versioned map) - if we want to change the parsing API for `PublicKeyCredentialSource` this should be discussed in a separate pull request.

Reply:
> My comments have nothing to do with representation. The representation is the same (some CBOR encoding).

My answer was about the fact that you say the representation is "some CBOR encoding". As far as I understand, the other structs in this file have a CBOR representation specified by FIDO. Is that correct?

If it is correct, then it would be a good design decision to not use the same API for representations that are specified by FIDO from the representations that are up to us. This would avoid confusion by having a consistent design. But maybe your point is that we don't care. Is that the case?

Also, note that using CBOR for credentials is accidental (since not specified). If we ever decide to serialize in a different way, it would break the API. Note also that we would actually need to do the same for the encoding depending on the decision.

> In any case, PublicKeyCredentialSource was already implementing TryFrom<cbor::Value>, so changing this API is independent from the core change of this pull request (using a map instead of an array).
> So this pull request should focus on addressing one change (the versioned map) - if we want to change the parsing API for PublicKeyCredentialSource this should be discussed in a separate pull request.

Indeed, I can make a first PR doing just this change. It would be cleaner. I didn't think it would be an issue.

Note that there is also a technical reason to change the API for credentials (beside the design reason), which is to avoid the copy of unknown fields. But this is minor since we expect those to be often empty. However we would still need a constant set of all valid fields, to be able to build the unknown fields by filtering the map.

Reply:
cc @kaczmarczyck

Reply:
I don't think we need to have different traits for CBOR for CTAP-defined data structures coming from USB and self-defined data structures going to our storage.
PublicKeyCredentialSource is in our control entirely and we can always rename it. As long as we are close to the WebAuthn version, I am happy with the current name.
I vote for TryFrom for easier code and make up for the lack of self-documenting with a comment.

Reply:
Done.